app/uk/gov/hmrc/plasticpackagingtaxregistration/connectors/SubscriptionsConnector.scala

```
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2021 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.plasticpackagingtaxregistration.connectors
+
+import java.util.UUID
+
+import com.kenshoo.play.metrics.Metrics
+import javax.inject.{Inject, Singleton}
+import uk.gov.hmrc.http.HttpReads.Implicits._
+import uk.gov.hmrc.http.{HeaderCarrier, HttpClient}
+import uk.gov.hmrc.plasticpackagingtaxregistration.config.AppConfig
+import uk.gov.hmrc.plasticpackagingtaxregistration.connectors.models.SubscriptionStatusResponse
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+@Singleton
+class SubscriptionsConnector @Inject() (
+  httpClient: HttpClient,
+  override val appConfig: AppConfig,
+  metrics: Metrics
+)(implicit ec: ExecutionContext)
+    extends EISConnector {
+
+  def getSubscriptionStatus(
+    safeNumber: String
+  )(implicit hc: HeaderCarrier): Future[SubscriptionStatusResponse] = {
+    val timer               = metrics.defaultRegistry.timer("ppt.subscription.status.timer").time()
+    val correlationIdHeader = correlationId -> UUID.randomUUID().toString
+    httpClient.GET[SubscriptionStatusResponse](appConfig.subscriptionStatusUrl(safeNumber),
+                                               headers = headers :+ correlationIdHeader
+    )
+      .andThen { case _ => timer.stop() }
+      .andThen {
+        case Success(response) => response
```

this is nice, didn't know you can do that, what happens when malformed response comes back? cannot serialise?

---

Yes, if it cannot serialiase to the Type, then a failure is returned.

I was reading about Async operations and how Scala is trying to move to Success and Failure and leave the Either[Left, Right] as it is not as intuitive and there is a bit of faf around.
Either still has a place, but in this case we just want to throw an exception as the message will not be useful to the client and customer.